lib/3scale/backend/stats/cleaner.rb

```diff
@@ -0,0 +1,156 @@
+module ThreeScale
+  module Backend
+    module Stats
+      class Cleaner
+        # Design notes:
+        # Apisonator does not store in any Redis structure the stats keys
+        # associated with a service. Doing so would imply:
+        #   - Performance hit when reporting. After updating a stats key, it
+        #   would need to be included in a set, which would increase the number
+        #   of operations in Redis.
+        #   - More space in Redis. To maintain the sets mentioned in the point
+        #   above.
+        #   - A data migration would be needed to create those sets from the
+        #   existing stats keys.
+        #
+        # In order to avoid those costs, this class is implemented in a way that
+        # does not need to keep an updated list of all the stats keys for every
+        # service. Services marked for deletion are stored in a Redis set, and
+        # then, a script, periodic cron, etc. is responsible for calling
+        # delete!(). That method scans the whole database and deletes all the
+        # stats keys that belong to those services marked to be deleted.
+        # The downside of this method is that it requires direct access to the
+        # redis servers. Going through a proxy like Twemproxy does not work,
+        # because it does not support the "SCAN" command.
+        #
+        # In the past we tried an alternative approach. When we received a
+        # request to delete the stats of a service, we generated all the
+        # possible stats keys that could exist for it. That approach was not
+        # efficient because it ended up generating many keys that didn't exist
+        # and thus, unnecessary delete calls to Redis. That approach is also
+        # more complex and error prone. You can find the details here:
+        # https://github.com/3scale/apisonator/issues/90
+
+        include Storable
+
+        SET_WITH_SERVICES_MARKED_FOR_DELETION = 'set_with_services_marked_for_deletion'.freeze
+        private_constant :SET_WITH_SERVICES_MARKED_FOR_DELETION
+
+        SLEEP_BETWEEN_SCANS = 0.01 # In seconds
```

Comment:
Make these a default fallback and allow users to specify them (sleep, slice size, ...)?

Reply:
The defaults here are similar to the ones that we use in the rest of the codebase. I don't know if moving this responsibility to the user is the right decision.

Reply:
They are fallbacks, to align to what's used elsewhere but giving the user the opt-in possibility (rather than the forced responsibility) of overriding them. Not a blocker.

Reply:
I propose to address this in a future PR.
First, let's see how this behaves in a real use case scenario. Let's see how often we get requests to delete stats, how long it takes for each shard, etc, and then come back to this, adjust those params if needed, and offer some options to tune them. What do you think?